### ISSUE TITLE
The backend no longer returns straight markdown.  Instead it is a json structure that looks like this.

### ISSUE THREAD

**Issue by binaryninja** [binaryninja/arxa#1]
The backend no longer returns straight markdown.  Instead it is a json structure that looks like this.  

We need to detect if the response is coming from the api server and if so run a new function that converts it into markdown.

Here is an exmaple if the data structure.  It should handle stuff being pretty dynamic apart from the first two levels of json.

{"research_notes": {"info": {"title": "Text Embeddings by Weakly-Supervised Contrastive Pre-training", "authors": ["Liang Wang", "Nan Yang", "Xiaolong Huang", "Binxing Jiao", "Linjun Yang", "Daxin Jiang", "Rangan Majumder", "Furu Wei"], "arxiv_link": "http://arxiv.org/abs/2212.03533v2",       
"date_of_submission": "22 Feb 2024", "field_of_study": "Natural Language Processing / Information Retrieval", "keywords": ["text embeddings", "contrastive learning", "weak supervision", "retrieval", "representation learning", "zero-shot", "fine-tuning", "BEIR", "MTEB"], "code_repository":     
"https://github.com/microsoft/unilm"}, "summary": {"problem_statement": "Traditional text embedding methods often require large amounts of labeled data or rely on synthetic signals that are of low quality, making it challenging to obtain robust and generalizable embeddings for retrieval,      
clustering, and classification tasks.", "main_contributions": ["Introduction of E5, a family of state-of-the-art text embeddings learned via weakly-supervised contrastive pre-training.", "Construction and use of a large-scale, curated text pair dataset (CCPairs) using heterogeneous web sources
and a consistency-based filtering mechanism.", "Demonstration that E5 can outperform standard baselines like BM25 in zero-shot settings and achieve competitive or superior results on the MTEB benchmark even with significantly fewer parameters than existing models."], "key_findings": ["E5-PT   
base outperforms BM25 on the BEIR benchmark without any labeled data in zero-shot retrieval settings.", "Fine-tuned E5 models achieve state-of-the-art performance on the MTEB benchmark, competing against models with 40\u00d7 more parameters.", "Data filtering and the design of contrastive     
pre-training with in-batch negatives are critical for model performance."], "methodology_overview": "The approach involves contrastive pre-training with in-batch negatives on a large curated dataset (CCPairs) consisting of diverse text pair signals, followed by supervised fine-tuning on       
smaller high-quality labeled datasets (e.g., MS-MARCO, NQ, NLI) to further refine the representations.", "conclusion": "E5 offers a strong general-purpose embedding model that transfers well across various tasks, demonstrating the importance of data quality and tailored pre-training strategies
in achieving superior performance compared to both traditional and contemporary approaches."}, "background": {"prior_work_referenced": "The work builds upon prior research in generating text embeddings using methods such as Sentence-BERT, SimCSE, as well as traditional techniques like TF-IDF  
and latent semantic analysis. It also references approaches using synthetic data (e.g., ICT, random cropping) and previous contrastive learning frameworks such as SimCLR.", "how_it_differs": "Unlike previous methods that rely either on limited labeled datasets or low-quality synthetic pairs,  
the paper carefully curates a large and diverse text pair dataset and employs a consistency-based filtering mechanism to ensure higher quality supervision. This enables the model to perform strongly in both unsupervised and supervised settings.", "gaps_addressed": "The paper addresses the     
challenge of obtaining robust text embeddings that can generalize across tasks without relying on massive labeled datasets, and it demonstrates that high-quality, weakly-supervised signals coupled with contrastive learning can outperform traditional IR baselines like BM25."}, "methodology":   
{"approach_taken": "A two-stage training process is adopted: contrastive pre-training on the CCPairs dataset using in-batch negatives, followed by supervised fine-tuning on diverse labeled datasets to inject specific human knowledge.", "key_techniques_used": ["Contrastive learning using       
InfoNCE loss", "In-batch negative sampling", "Consistency-based data filtering", "Fine-tuning with supervised signals and knowledge distillation from a cross-encoder teacher model"], "datasets_benchmarks_used": "A large-scale text pair dataset (CCPairs) assembled from Reddit, Common Crawl,    
StackExchange, Wikipedia, scientific papers, and news; Evaluation is performed on the BEIR benchmark (15 datasets) and the MTEB benchmark (56 datasets spanning multiple tasks).", "implementation_details": "Models are based on BERT architectures (E5 small, base, and large) with careful         
hyperparameter tuning such as using large batch sizes (up to 32K for pre-training) and employing mixed precision training with gradient checkpointing. Fine-tuning involves multi-dataset training with hard negative mining and a combination of contrastive loss and KL divergence for              
distillation.", "reproducibility": "The authors provide implementation details, hyperparameter settings, and a link to the public code repository, ensuring that the experiments can be reproduced with access to the described datasets and training configurations."}, "evaluation":                
{"evaluation_metrics": "Metrics such as nDCG@10, MRR@10, Recall, accuracy, v-measure, MAP, and Spearman rank correlation are used. Both domain-specific evaluations (e.g., MS-MARCO passage ranking) and cross-task metrics from BEIR and MTEB are reported.", "results_summary": "E5 models, whether 
used in an unsupervised or fine-tuned setting, consistently outperform baseline methods including BM25, SimCSE, and other dense retrieval models. Notably, the E5-PT base model beats BM25 in zero-shot settings on BEIR and fine-tuned versions achieve higher scores on extensive evaluations in    
MTEB.", "baseline_comparisons": "Comparisons are made against classical baselines like BM25 and other modern models (SimCSE, Contriever, GTR, Sentence-T5). The results highlight that despite having fewer parameters than some competitors, E5 models achieve competitive or improved performance.",
"ablation_studies": "Ablation studies include analysis of the impacts of batch size, negative sampling strategies, and data filtering techniques on performance. These studies validate the key role of each component (e.g., larger batch sizes and consistency-based filtering) in boosting the     
model\u2019s effectiveness.", "limitations_noted": "While the model performs strongly in many settings, limitations remain in densely retrieving from domains such as long documents or in cases where exact lexical matches are critical. Further, the reliance on a complex data curation process   
could pose challenges in other contexts."}, "strengths": {"novelty_innovation": "The introduction of a large, carefully curated weakly-supervised dataset (CCPairs) combined with a simple yet effective contrastive training paradigm is highly innovative.", "technical_soundness": "The method is  
backed by thorough experiments, ablation studies, and comparisons across diverse benchmarks, demonstrating the robustness of the approach.", "clarity_and_organization": "The paper is well-structured, clearly outlining the motivation, methodology, experimental design, and findings.",           
"impact_applications": "E5 shows significant promise for practical applications in retrieval, clustering, and classification, potentially influencing future designs in text embedding and dense retrieval systems."}, "weaknesses": {"unaddressed_assumptions_flaws": "The approach assumes that a   
large amount of weakly supervised data and aggressive filtering will generalize across all domains, which might not hold in every scenario.", "possible_biases_limitations": "There is potential domain bias given the sources of the text pairs (e.g., web data, Reddit, scientific papers), and some
tasks like long-document retrieval may still benefit from traditional methods such as BM25.", "reproducibility_concerns": "While details are provided, the data curation and filtering pipeline is complex and may be challenging to replicate fully without access to identical resources.",         
"presentation_issues": "Certain experimental details, particularly regarding the tuning of negative sampling strategies and the iterative filtering process, could be elaborated upon for clarity."}, "future_work": {"suggested_improvements": "Exploration of iterative filtering techniques,       
further refinement of negative sampling strategies, and adaptation of the model to domains with long textual content.", "potential_extensions": "Integrating multi-modal signals, scaling up model parameters, and exploring multilingual embeddings using similar weakly supervised contrastive      
methods.", "open_problems": "Addressing the gap between dense and sparse retrieval for long-tail domains and improving robustness in settings requiring exact lexical matches remain open challenges."}, "personal_review": {"overall_impression": "The paper presents a well-motivated and           
methodologically sound approach to text embeddings. Its innovation in data curation and contrastive pre-training provides a valuable contribution to the field.", "significance_contributions": "The introduction of the E5 model family marks a meaningful advance in general-purpose text embedding 
techniques, especially given its competitive performance with a fraction of the parameters of some larger models.", "clarity_and_organization": "The document is clearly written and logically structured, making it accessible to researchers and practitioners alike.", "methodological_rigor": "The
extensive evaluations across multiple benchmarks and the inclusion of ablation studies demonstrate high methodological rigor.", "reproducibility": "While the availability of code and detailed hyperparameters are strengths, the complexity of the data filtering process may pose some challenges  
for complete reproducibility."}, "additional_notes": {"key_takeaways": "High-quality, weakly supervised data combined with contrastive pre-training can produce robust text embeddings that excel in both unsupervised and supervised scenarios.", "interesting_insights": "The study highlights the  
effectiveness of a consistency-based filtering mechanism to improve training data quality and demonstrates that careful dataset curation can lead to significant performance gains.", "personal_thoughts": "This work sets a strong precedent for future research in text embeddings, and its         
practical implications in retrieval and classification systems are particularly noteworthy."}}}
